hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java

```diff
@@ -287,21 +290,44 @@ public Option<HoodieIndexMetadata> getIndexMetadata() {
     if (indexMetadataOpt.isPresent() && !indexMetadataOpt.get().getIndexDefinitions().isEmpty()) {
       return indexMetadataOpt;
     }
+    Option<HoodieIndexMetadata> indexDefOption = Option.empty();
     if (tableConfig.getRelativeIndexDefinitionPath().isPresent() && StringUtils.nonEmpty(tableConfig.getRelativeIndexDefinitionPath().get())) {
-      StoragePath indexDefinitionPath =
-          new StoragePath(basePath, tableConfig.getRelativeIndexDefinitionPath().get());
-      try {
-        Option<byte[]> bytesOpt = FileIOUtils.readDataFromPath(storage, indexDefinitionPath, true);
-        if (bytesOpt.isPresent()) {
-          return Option.of(HoodieIndexMetadata.fromJson(new String(bytesOpt.get())));
-        } else {
-          return Option.of(new HoodieIndexMetadata());
-        }
-      } catch (IOException e) {
-        throw new HoodieIOException("Could not load index definition at path: " + tableConfig.getRelativeIndexDefinitionPath().get(), e);
+      indexDefOption = loadIndexDefFromStorage(basePath, tableConfig.getRelativeIndexDefinitionPath().get(), storage);
+    }
+    populateIndexVersionIfMissing(tableConfig.getTableVersion(), indexDefOption);
+    return indexDefOption;
+  }
+
+  public static void populateIndexVersionIfMissing(HoodieTableVersion tableVersion, Option<HoodieIndexMetadata> indexDefOption) {
+    indexDefOption.ifPresent(idxDefs ->
+        idxDefs.getIndexDefinitions().replaceAll((indexName, idxDef) -> {
+          ValidationUtils.checkArgument(HoodieIndexVersion.isValidIndexDefinition(tableVersion, idxDef),
+              String.format("Table version %s, index definition %s", tableVersion, idxDef));
+          if (idxDef.getVersion() == null) {
+            // If version field is missing, it implies either of the cases (validated by isValidIndexDefinition):
+            // - It is table version 8, because we don't write version attributes in some hudi releases
+            // - It is table version 9, and it is not secondary index. Since we drop SI on upgrade and we always write version attributes.
+            return idxDef.toBuilder().withVersion(getCurrentVersion(tableVersion, idxDef.getIndexName())).build();
```

Comment:
lets add tests to validate the index defs are written with valid index versions on storage.. (if we don't have em). 

instead of all this - can we just write index version as a part of upgrade to table version 9? i.e read and one time rewrite the index defs.. 

Then the regular path becomes easy : if there is no index def version i.e. null (only possible due to a table version < 9 read by table version >= 9 process), then we assume v1..

This vastly simplifies ongoing maintenance.